pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/MultiStageBrokerRequestHandler.java

```diff
@@ -224,67 +227,89 @@ protected BrokerResponse handleRequest(long requestId, String query, SqlNodeAndO
       return new BrokerResponseNative(QueryException.getException(QueryException.QUOTA_EXCEEDED_ERROR, errorMessage));
     }
 
-    Tracing.ThreadAccountantOps.setupRunner(String.valueOf(requestId), ThreadExecutionContext.TaskType.MSE);
-
-    long executionStartTimeNs = System.nanoTime();
-    QueryDispatcher.QueryResult queryResults;
+    Timer queryTimer = new Timer(queryTimeoutMs);
     try {
-      queryResults =
-          _queryDispatcher.submitAndReduce(requestContext, dispatchableSubPlan, queryTimeoutMs, queryOptions);
-    } catch (TimeoutException e) {
-      for (String table : tableNames) {
-        _brokerMetrics.addMeteredTableValue(table, BrokerMeter.BROKER_RESPONSES_WITH_TIMEOUTS, 1);
+      // It's fine to block in this thread because we use a separate thread pool from the main Jersey server to process
+      // these requests.
+      if (!_querySemaphore.tryAcquire(queryTimeoutMs, TimeUnit.MILLISECONDS)) {
+        LOGGER.warn("Timed out waiting to execute request {}: {}", requestId, query);
+        requestContext.setErrorCode(QueryException.EXECUTION_TIMEOUT_ERROR_CODE);
+        return new BrokerResponseNative(QueryException.EXECUTION_TIMEOUT_ERROR);
       }
-      LOGGER.warn("Timed out executing request {}: {}", requestId, query);
+    } catch (InterruptedException e) {
+      LOGGER.warn("Interrupt received while waiting to execute request {}: {}", requestId, query);
       requestContext.setErrorCode(QueryException.EXECUTION_TIMEOUT_ERROR_CODE);
       return new BrokerResponseNative(QueryException.EXECUTION_TIMEOUT_ERROR);
```

Comment:
Maybe not in this PR, but we will probably need to have a way to tell the user where we were waiting most of the time. It could be useful for example to tell users that their queries are timing out because other queries are consuming all threads.

Reply:
a metric would be great, a log could be useful. But I was more worried about immediate the feedback to the user. I mean, the error information we return. But it was a wishful thinking more than an actual requirement.